Support for multiple managed worlds

QuantumToasted opened the pull request:
As I mentioned in #77, I thought it would be nice for PopulationDensity to support multiple managed worlds as opposed to only a single world. I know that such a big undertaking would be a large amount of work and might require lots of testing and many sets of eyes to observe the changes to code to make sure nothing would break, and everything would work the right way.

I am opening this PR with the rough changes I have so far, with the hopes that we can collectively get the ball rolling for this type of support!

Notable code changes:

- `PopulationDensity.ManagedWorld` has become `PopulationDensity.ManagedWorlds` and is now of type `List<World>`. I did my best to ensure any references or checks for `ManagedWorld` would now perform the proper reference or check inside `ManagedWorlds` (that is, to ensure that the player does not do something in one managed world that would affect another). Testing will be needed to ensure this is the case.
- `RegionCoordinates` now also contains a `world` property/field such that methods which require it can now be linked to the world they're for. It prevents needing to pass `World`s into every method, but if you do not like this, it can be removed. I also modified the `toString()`, `public RegionCoordinates(String string)`, and `equals()` methods to better accomodate and reflect these changes.
- A few small code changes here and there were made to better communicate that there can be multiple managed worlds. Some console logging makes note to add world info during region scans.
- DataStore has a small change, namely that region data files now are in the format `world,x z` instead of the previous `x z` format, to ensure regions are loaded for the right world. (See changes to `RegionCoordinates` for more info).
- Many others I might have missed

Do note that much of this code was written or hacked at in the wee morning hours as I was determined to create a working build before sleeping. I do apologies for any gross coding styles, and I greatly appreciate and welcome any changes to better suit your code style. I come from a C# background so I may have missed some common Java style choices.

The one caveat to all of this, however, is that it means that users will need to regenerate or modify their `config.yml` file to reflect this multiple-world change, otherwise their original managed world may not be detected. I'm sure that could simply be changed by adding a few lines of code to support either single or multiple worlds. It's up to you :)

There is a lot of changes here, so please do not hesitate to fix things or ask me why I did things the way I did. I'll answer to the best of my ability.

QuantumToasted committed:
Initial support for multiple managed worlds

Very rough work, a fair amount of replacing here and there, but at the very least, it does not cause console errors or warnings save for server slowdowns or long periods of inactivity due to resource scans.

RoboMWM commented on src/me/ryanhamshire/PopulationDensity/DataStore.java:
```diff
@@ -46,14 +42,7 @@
-import java.util.List;
-import java.util.Locale;
-import java.util.Random;
+import java.util.*;
```
no wildcard imports pls

QuantumToasted replied:
Can probably do. I don't believe I ever manually added imports, so I blame IDEA for this. I'll nuke this line in the IDE and see which of the classes or whatever made it think this was necessary. :+1:

RoboMWM replied:
Still not resolved, fyi.

RoboMWM replied:
But ya, intellij IDEA does this by default. Can increase default 5 to some large number like 99

QuantumToasted replied:
I finally fixed it, changing that setting to 99 seemed to help. :p

RoboMWM commented on src/me/ryanhamshire/PopulationDensity/DataStore.java:
```diff
@@ -71,15 +60,15 @@
 
     //coordinates of the next region which will be opened, if one needs to be opened
-    private RegionCoordinates nextRegionCoordinates;
+    private ConcurrentHashMap<World, RegionCoordinates> nextRegionCoordinates = new ConcurrentHashMap<>();
```
Use Map interface instead of implementation?

QuantumToasted replied:
Is this about the left-hand side of the `=`? I swear for the right-hand side, IDEA said that you can't do `new Map<>()` because it's an interface, much like in C#.
```suggestion
    private Map<World, RegionCoordinates> nextRegionCoordinates = new ConcurrentHashMap<>();
```

RoboMWM replied:
Code suggestions? Wow I must've missed some GitHub blog update on PRs...

But ya, that.

QuantumToasted committed:
Replaced wildcard imports with dedicated imports

QuantumToasted committed:
Replaced concrete ArrayList declaration with List

QuantumToasted committed:
Replaced an == with equals()

I have not bothered to dig through the source code for Bukkit/Spigot, so I am unaware if == is equivalent to .equals() in this case.

QuantumToasted committed:
Replaced concrete ConcurrentHashMap declaration with Map

QuantumToasted committed:
Replaced inline brackets with newline brackets

Tried CTRL+SHIFT+F in IDEA, let me know if there are any i missed.

QuantumToasted committed:
Merge remote-tracking branch 'origin/master'

QuantumToasted committed:
Replaced (another) concrete ConcurrentHashMap declaration with Map

For some reason, IDEA's building tools let this slip by, and caused it to build successfully but fail to compile. I don't get it.

QuantumToasted committed:
Removed wildcard imports

Hopefully for the last damn time.

QuantumToasted committed:
Another interface declaration, List<> instead of ArrayList<>

I don't remember who originally wrote this line of code, if it wasn't me, I just saw it while i was fixing imports so I thought to be consistent.
